README.md
R package: queryup
================
Guillaume Voisinne
2023 - 02 - 08

[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/queryup)](https://cran.r-project.org/package=queryup)
[![R-CMD-check](https://github.com/VoisinneG/queryup/workflows/R-CMD-check/badge.svg)](https://github.com/VoisinneG/queryup/actions)
[![Codecov test
coverage](https://codecov.io/gh/VoisinneG/queryup/branch/master/graph/badge.svg)](https://app.codecov.io/gh/VoisinneG/queryup?branch=master)
[![CRAN mirror
downloads](https://cranlogs.r-pkg.org/badges/queryup)](https://cran.r-project.org/package=queryup/)

The `queryup` R package aims to facilitate retrieving information from
the UniProt database using R. Programmatic access to the UniProt
database is performed by submitting queries to the [UniProt website REST
API](https://www.uniprot.org/help/api_queries).

## Install

You can install the package from CRAN using:

``` r
install.packages("queryup")
```

Alternatively, you may also install the package from github using
devtools:

``` r
devtools::install_github("VoisinneG/queryup")
```

## Queries

Queries combine different fields to identify matching database entries.
Here, queries are submitted using the function `query_uniprot()`. In the
`queryup` R package, a query must be formatted as a list containing
character vectors named after existing UniProt fields (available query
fields can be found in the [API
documentation](https://www.uniprot.org/help/query-fields) or in the
package data `query_fields$field`). Different query fields must be
matched simultaneously. For instance, the following query uses the
fields *gene_exact* to return the UniProt entries of all proteins
encoded by gene *Pik3r1* :

``` r
library(queryup)
```

``` r
query <- list("gene_exact" = "Pik3r1")
df <- query_uniprot(query, show_progress = FALSE)
head(df)
```

    ##        Entry       Entry Name Gene Names Organism (ID)   Reviewed
    ## 2 A0A096MNU6 A0A096MNU6_PAPAN     PIK3R1          9555 unreviewed
    ## 3 A0A0D9RTM6 A0A0D9RTM6_CHLSB     PIK3R1         60711 unreviewed
    ## 4 A0A1S3F3Z7 A0A1S3F3Z7_DIPOR     Pik3r1         10020 unreviewed
    ## 5 A0A1U7Q814 A0A1U7Q814_MESAU     Pik3r1         10036 unreviewed
    ## 6 A0A287DCB8 A0A287DCB8_ICTTR     PIK3R1         43179 unreviewed
    ## 7 A0A2I2ZTD7 A0A2I2ZTD7_GORGO     PIK3R1          9595 unreviewed

Available query fields can be listed using the package data
`query_fields`:

``` r
query_fields$field
```

    ##  [1] "accession"                                                
    ##  [2] "active"                                                   
    ##  [3] "Refer to the page: Sequence Annotations"                  
    ##  [4] "lit_author"                                               
    ##  [5] "protein_name"                                             
    ##  [6] "chebi"                                                    
    ##  [7] "uniprot_id (/uniref), then uniref_cluster_90 (/uniprotkb)"
    ##  [8] "xrefcount_pdb (or xref_count)"                            
    ##  [9] "date_created"                                             
    ## [10] "database, xref"                                           
    ## [11] "ec"                                                       
    ## [12] "Refer to the pages: Comments or Sequence Annotations"     
    ## [13] "existence"                                                
    ## [14] "family"                                                   
    ## [15] "fragment"                                                 
    ## [16] "gene"                                                     
    ## [17] "gene_exact"                                               
    ## [18] "go"                                                       
    ## [19] "virus_host_name, virus_host_id"                           
    ## [20] "accession_id"                                             
    ## [21] "inchikey"                                                 
    ## [22] "protein_name"                                             
    ## [23] "interactor"                                               
    ## [24] "keyword"                                                  
    ## [25] "length"                                                   
    ## [26] "mass"                                                     
    ## [27] "cc_mass_spectrometry"                                     
    ## [28] "date_modified"                                            
    ## [29] "protein_name"                                             
    ## [30] "organelle"                                                
    ## [31] "organism_name, organism_id"                               
    ## [32] "plasmid"                                                  
    ## [33] "proteome"                                                 
    ## [34] "proteomecomponent"                                        
    ## [35] "sec_acc"                                                  
    ## [36] "reviewed"                                                 
    ## [37] "scope"                                                    
    ## [38] "sec_acc"                                                  
    ## [39] "sequence"                                                 
    ## [40] "date_sequence_modified"                                   
    ## [41] "strain"                                                   
    ## [42] "taxonomy_name, taxonomy_id"                               
    ## [43] "tissue"                                                   
    ## [44] "cc_webresource"

## Columns

By default, `query_uniprot()` returns a data.frame with UniProt
accession IDs, gene names, organism and Swiss-Prot review status. You
can choose which data columns to retrieve using the `columns` parameter.

``` r
df <- query_uniprot(query, 
                    columns = c("id", "sequence", "keyword", "gene_primary"),
                    show_progress = FALSE)
```

    ## Warning in (function (..., deparse.level = 1) : number of columns of result is
    ## not a multiple of vector length (arg 881)

See the [API documentation](https://www.uniprot.org/help/return_fields)
or the package data `return_fields` for all available columns. Available
returned fields can be listed using the package data `return_fields`:

``` r
head(return_fields)
```

    ##          field                      label
    ## 1    accession                      Entry
    ## 2           id                 Entry name
    ## 3   gene_names                 Gene names
    ## 4 gene_primary       Gene names (primary)
    ## 5 gene_synonym       Gene names (synonym)
    ## 6     gene_oln Gene names (ordered locus)

Note that the parameter `columns` and the name of the corresponding
column in the output data frame do not necessarily match (they
correspond to columns “field” and “label” respectively in the package
data `return_fields`).

``` r
names(df)
```

    ## [1] "Entry"                "Entry Name"           "Sequence"            
    ## [4] "Keywords"             "Gene Names (primary)"

Let’s check the sequence and the UniProt keywords corresponding to the
first entry :

``` r
as.character(df$Sequence[1])
```

    ## [1] "MSAEGYQYRALYDYKKEREEDIDLHLGDILTVNKGSLVALGFSDGQEARPEEIGWLNGYNETTGERGDFPGTYVEYIGRKKISPPTPKPRPPRPLPVAPGSSKTEADVEQQALTLPDLAEQFAPPDVAPPLLIKLVEAIEKKGLECSTLYRTQSSGNLAELRQLLDCDTASVDLEMIDVHILADAFKRYLLDLPNPVIPAAVYSEMISLAQEVQSSEEYIQLLKKLIRSPSIPHQYWLTLQYLLKHFFKLSQTSSKNLLNARVLSEIFSPMLFRFSAASSDNTENLIKVIEILISTEWNERQPAPALPPKPPKPTTVANNGMNNNMSLQDAEWYWGDISREEVNEKLRDTADGTFLVRDASTKMHGDYTLTLRKGGNNKLIKIFHRDGKYGFSDPLTFNSVVELINHYRNESLAQYNPKLDVKLLYPVSKYQQDQVVKEDNIEAVGKKLHEYNTQFQEKSREYDRLYEEYTRTSQEIQMKRTAIEAFNETIKIFEEQCQTQERYSKEYIEKFKREGNEKEIQRIMHNYDKLKSRISEIIDSRRRLEEDLKKQAAEYREIDKRMNSIKPDLIQLRKTRDQYLMWLTQKGVRQKKLNEWLGNENTEDQYSLVEDDEDLPHHDEKTWNVGSSNRNKAENLLRGKRDGTFLVRESSKQGCYACSVVVDGEVKHCVINKTATGYGFAEPYNLYSSLKELVLHYQHTSLVQHNDSLNVTLAYPVYAQDSYFIFQGNMGRMHGNGHSM"

``` r
as.character(df$Keywords[1])
```

    ## [1] "Coiled coil;Protein transport;Reference proteome;Repeat;SH2 domain;SH3 domain;Stress response;Transport"

## Combining query fields

Our first query returned many matches. We can build more specific
queries by using more than one query field. By default, matching entries
must satisfy all query fields simultaneously. Let’s retrieve the only
Swiss-Prot reviewed protein entry encoded by gene *Pik3r1* in *Homo
sapiens* (taxon: 9606):

``` r
query <- list("gene_exact" = "Pik3r1", 
              "reviewed" = "true", 
              "organism_id" = "9606")
df <- query_uniprot(query, show_progress = FALSE)
print(df)
```

    ##    Entry Entry Name  Gene Names Organism (ID) Reviewed
    ## 2 P27986 P85A_HUMAN PIK3R1 GRB1          9606 reviewed

## Multiple items per query field

It is also possible to look for entries that match different items
within a single query field. Items from a given query field are looked
for independently. Hence, the following query will return all Swiss-Prot
reviewed proteins encoded by either *Pik3r1* or *Pik3r2* in either *Mus
musculus* (taxon: 10090) or *Homo sapiens* (taxon: 9606):

``` r
query <- list("gene_exact" = c("Pik3r1", "Pik3r2"), 
              "reviewed" = "true", 
              "organism_id" = c("9606", "10090"))
df <- query_uniprot(query, show_progress = FALSE)
print(df)
```

    ##    Entry Entry Name  Gene Names Organism (ID) Reviewed
    ## 2 O00459 P85B_HUMAN      PIK3R2          9606 reviewed
    ## 3 O08908 P85B_MOUSE      Pik3r2         10090 reviewed
    ## 4 P26450 P85A_MOUSE      Pik3r1         10090 reviewed
    ## 5 P27986 P85A_HUMAN PIK3R1 GRB1          9606 reviewed

## Queries with invalid entries

If a query containing invalid entries is sent to the UniProt REST API,
an error message is returned and no information about the other
potentially valid entries can be retrieved. To overcome this limitation,
`queryup` parses the error messages and remove invalid entries from the
query. Hence, `query_uniprot()` will return information for valid
entries only :

``` r
invalid_ids <- c("P226", "CON_P22682", "REV_P47941")
valid_ids <- c("A0A0U1ZFN5", "P22682")
ids <- c(invalid_ids, valid_ids)
query <- list("accession_id" = ids)
query_uniprot(query)
```

    ## 3 invalid values were found (P226, CON_P22682, REV_P47941) and removed from the query.

    ##        Entry     Entry Name Gene Names Organism (ID)   Reviewed
    ## 2 A0A0U1ZFN5 A0A0U1ZFN5_RAT  Cbl c-Cbl         10116 unreviewed
    ## 3     P22682      CBL_MOUSE        Cbl         10090   reviewed

## Long queries

Because UniProt REST API limits the size of queries, long queries
containing more than a few hundreds entries cannot be passed in a single
request. To overcome this limitation, the `queryup` package splits long
queries into smaller ones. For instance, the dataset `uniprot_entries`
that is bundled with the `queryup` package contains information for 1000
UniProt entries. We could retrieve the ENSEMBL ids corresponding to
these entries using :

``` r
ids <- uniprot_entries$Entry
query <- list("accession_id" = ids)
columns <- c("gene_names", "xref_ensembl")
df <- query_uniprot(query, columns = columns, show_progress = FALSE)
head(df)
```

    ##        Entry                 Gene Names
    ## 2 A0A087WPF7             Auts2 Kiaa0442
    ## 3 A0A088MLT8 Iqcj-Schip1 Iqschfp Schip1
    ## 4 A0A0B4J1F4                     Arrdc4
    ## 5 A0A0B4J1G0               Fcgr4 Fcgr3a
    ## 6 A0A0G2JDV3                 Gbp6 Mpa2l
    ## 7 A0A0U1RPR8                     Gucy2d
    ##                                                                Ensembl
    ## 2 ENSMUST00000161226 [A0A087WPF7-1];ENSMUST00000161374 [A0A087WPF7-3];
    ## 3                                   ENSMUST00000182006 [A0A088MLT8-1];
    ## 4 ENSMUST00000048068 [A0A0B4J1F4-1];ENSMUST00000118110 [A0A0B4J1F4-2];
    ## 5                                                  ENSMUST00000078825;
    ## 6                                                           A0A0G2JDV3
    ## 7                                                  ENSMUST00000206435;

## Protein-protein interactions

Another usage could be to retrieve protein-protein interactions among a
set of UniProt entries:

``` r
ids <- sample(uniprot_entries$Entry, 400)
query <- list("accession_id" = ids, 
              "interactor" = ids)
columns <- "cc_interaction"
df <- query_uniprot(query = query, columns = columns, show_progress = FALSE)
head(df)
```

    ##     Entry                                         Interacts with
    ## 2  A2A259                                         Q2EG98; A2A259
    ## 3  O88273                                                 O88273
    ## 4  O88522 Q60680; O88351; O88522; Q924T7; P62991; P0CG48; P24772
    ## 21 E9Q401                 Q6PHZ2; Q9Z2I2; Q8K4S1; E9Q401; P23327


NEWS.md
# queryup 1.0.5

* Handling of possible errors when using `httr::http_type()` for long queries

# queryup 1.0.4

* Handling of possible errors when using `httr::content()`

# queryup 1.0.3

* Better handling of possible errors when using `httr::GET()`

# queryup 1.0.2

* Package dependencies on `RCurl` and `jsonlite` packages have been replaced 
by a single dependency on the `httr` package.
* When a request fails, a message with the http status code is returned.
* Parameters `print_url` and `print_uniprot_messages` have been deprecated.

# queryup 1.0.1

* The function `get_uniprot_data` now works also for long queries (with more 
than 300 items) 

# queryup 1.0.0

* `queryup` is now compatible with the latest UniProt REST API (rest.uniprot.org)
* messages are more informative and will let you know invalid query fields,
invalid return fields and invalid entries
* `queryup` filters out invalid UniProt entries from queries
* `queryup` can manage long queries (with hundreds of entries)


tests/testthat.R
library(testthat)
library(queryup)

test_check("queryup")


tests/testthat/test-queryup.R
test_that("Only objects of type 'list' or 'character' vectors of size 1 are accepted as inputs", {
  expect_equal(get_uniprot_data(NULL), NULL)
  expect_equal(get_uniprot_data(), NULL)
  expect_equal(get_uniprot_data(1:10), NULL)
  expect_equal(get_uniprot_data(c("P22682", "P47941"), print_full_query = TRUE), NULL)
})

test_that("Basic query works", {

  existing_ids <- c("P22682", "P47941")
  proper_query <- list("accession_id" = existing_ids)
  res <- get_uniprot_data(proper_query)

  expect_equal(is(res, "data.frame"), TRUE)
  expect_equal("Entry" %in% names(res), TRUE)
  expect_equal(dim(res)[1], length(existing_ids))
  expect_equal(setequal(res[["Entry"]], existing_ids), TRUE)
})

test_that("Query with unrecognized keys return an error", {

  existing_ids <- c("CON_P22682", "CON_P47941")
  query = list("acc_id" = existing_ids)
  expect_error(get_uniprot_data(query, ))

})


tests/testthat/test-query_uniprot.R
test_that("Query with incorrect base url returns a message", {
  existing_ids <- c("P22682", "P47941")
  query <- list("accession_id" = existing_ids)
  expect_message(query_uniprot(query,
                               base_url = "https://rest.uniprot.org/uni/"))
})

test_that("Long queries work", {
  # query with 500 entries
  ids <- uniprot_entries$Entry[1:500]
  query <- list("accession_id" = ids)
  res <- query_uniprot(query)
  if(!is.null(res)){
    expect_true(inherits(res, "data.frame"))
  }
})

test_that("Long queries with no columns defined work", {
  # query with 500 entries
  ids <- uniprot_entries$Entry[1:500]
  query <- list("accession_id" = ids)
  res <- query_uniprot(query, columns = NULL)

  if(!is.null(res)){
    expect_true(inherits(res, "data.frame"))
  }
})


test_that("Short queries work", {
  # query with 10 entries
  ids <- uniprot_entries$Entry[1:10]
  query <- list("accession_id" = ids)
  res <- query_uniprot(query)

  if(!is.null(res)){
    expect_true(inherits(res, "data.frame"))
  }
})

test_that("Setting 'max_keys' above 200 produces a message", {
  expect_message(query_uniprot("P22682", max_keys = 301), "200")
})

test_that("Query works when multiple fields have more than 'max_keys=200' items", {
  # query looking for interactions between 400 different entries
  ids <- sample(uniprot_entries$Entry, 400)
  query <- list("accession_id" = ids, "interactor" = ids)
  res <- query_uniprot(query, max_keys = 200,
                       columns = c("accession", "cc_interaction"))

  if(!is.null(res)){
    expect_true(inherits(res, "data.frame"))
  }
})


R/data.R
#' Information for 1000 UniProt entries from the organism Mus musculus
#'
#' Entry names and other attributes of 1000 UniProt entries
#' in Mus musculus.
#'
#' @format A data frame with 1000 rows and 5 variables:
#' \describe{
#'   \item{Entry}{UniProt entry accession id}
#'   \item{Entry Name}{UniProt entry name}
#'   \item{Gene Names}{Gene names}
#'   \item{Organism (ID)}{Taxon ID}
#'   \item{Reviewed}{Swiss-Prot review status}
#' }
#' @source \url{https://www.uniprot.org/}
"uniprot_entries"

#' Available query fields.
#'
#' Query fields that can be used to generate queries using `queryup`
#' along with associated examples and description.
#'
#' @format A data frame with 44 rows and 3 variables:
#' \describe{
#'   \item{field}{Name of the query field}
#'   \item{example}{Example query (as appearing in the query url)}
#'   \item{description}{Description of the example query}
#' }
#' @source \url{https://www.uniprot.org/help/query-fields}
"query_fields"

#' Available return fields ("columns").
#'
#' Return fields that can be retrieved using `queryup`
#' along with their label (column "Label") as appearing in the retrieved
#' data.frame.
#'
#' @format A data frame with 287 rows and 2 variables:
#' \describe{
#'   \item{field}{Name of the returned field}
#'   \item{label}{Label of the corresponding column in the retrieved data.frame}
#' }
#' @source \url{https://www.uniprot.org/help/return_fields}
"return_fields"


R/queryu.R
#' Retrieve data from uniprot using uniprot's REST API
#'
#' @param query list of keys corresponding to uniprot's query fields. For example :
#' list("gene_exact" = c("Pik3r1", "Pik3r2") , "organism" = c("10090", "9606"), "reviewed" = "yes")
#' @param columns names of uniprot data columns to retrieve. Examples include "id",
#' "genes", "keywords", "sequence"
#'
#' @return a data.frame
#' @import utils
#' @export
#'
#' @examples
#' #Getting gene names, keywords and protein sequences associated with a set of uniprot IDs.
#' ids <- c("P22682", "P47941")
#' cols <- c("id", "genes", "keywords", "sequence")
#' df <- get_uniprot_data(query = list("id" = ids), columns = cols)
#'
#' #Lists all entries describing interactions with the protein described by entry P00520.
#' df <- get_uniprot_data(query = list("interactor" = "P00520"), columns = cols)
get_uniprot_data <- function(query = NULL, columns = c("id", "keywords")){

  df <- NULL

  if(!is.null(query)){

    formatted_queries <- sapply(1:length(query),
                                function(x){paste(names(query)[x], ":(",
                                                  paste(query[[x]], collapse = "+or+"), ")",
                                                  sep ="")})

    url <- 'https://www.uniprot.org/uniprot/?query='
    full_query <- paste(formatted_queries, collapse = "+and+")

    cols <- tolower(paste(columns, collapse = ","))
    full_url <- paste('https://www.uniprot.org/uniprot/?query=', full_query,
                      '&format=tab&columns=', cols,
                      sep = "")

    message(paste("Querying uniprot...\n",sep=""))

    df <- tryCatch({
      read.table(full_url,
                 sep ="\t",
                 header = TRUE,
                 quote = "")
    }, error=function(err) {
      message(
        "reading url",
        "\n    ", full_url,
        "\nfailed"
      )
      NULL
    })
  }


  return(df)

}


#' Retrieve data from uniprot using uniprot's REST API.
#' To avoid non-responsive queries, they are splitted into
#' smaller queries withat most \code{max_keys} items per query field.
#'
#' @param query list of keys corresponding to uniprot's query fields. For example :
#' list("gene_exact" = c("Pik3r1", "Pik3r2") , "organism" = c("10090", "9606"), "reviewed" = "yes")
#' @param columns names of uniprot data columns to retrieve. Examples include "id",
#' "genes", "keywords", "sequence"
#' @param max_keys maximum number of field items submitted
#' @return a data.frame
#' @export
query_uniprot <- function(query = NULL, columns = c("id", "genes", "organism", "reviewed" ), max_keys = 400 ){

  for ( i in 1:length(query)){

    if(length(query[[i]]) > max_keys){

      query_split <- vector("list", 2)

      query_split[[1]] <- query
      query_split[[2]] <- query

      query_split[[1]][[i]] <- query[[i]][1:max_keys]
      query_split[[2]][[i]] <- query[[i]][(max_keys+1):length(query[[i]])]

      #query_split[[1]] <- split_query(query_split[[1]], max_keys = max_keys)
      #query_split[[2]] <- split_query(query_split[[2]], max_keys = max_keys)

      df_list <- lapply(query_split, query_uniprot, columns = columns, max_keys = max_keys)

      return(do.call(rbind, df_list))

    }
  }

  return( get_uniprot_data(query, columns = columns) )

}


R/query_uniprot.R
#' Retrieve data from UniProt using UniProt's REST API.
#'
#' Retrieve data from UniProt using UniProt's REST API.
#' To avoid non-responsive queries, they are split into
#' smaller queries with at most \code{max_keys} items per query field.
#' Not that it works only with queries where items within query fields are
#' collapsed with '+OR+' and different
#' query fields are collapsed with '+AND+' (see \code{query_uniprot()})
#'
#' @param query list of keys corresponding to UniProt's query fields.
#' For example :
#' query = list("gene_exact" = c("Pik3r1", "Pik3r2"),
#' "organism_id" = c("10090", "9606"), "reviewed" = "true").
#' See `query_fields` for available query fields.
#' @param base_url The base url for the UniProt REST API
#' @param columns names of UniProt data columns to retrieve.
#' Examples include "accession", "id", "genes", "keywords", "sequence".
#' See `return_fields` for available return fields.
#' @param max_keys maximum number of field items submitted
#' @param updateProgress used to display progress in shiny apps
#' @param show_progress Show progress bar
#' @return a data.frame
#' @importFrom utils setTxtProgressBar txtProgressBar
#' @export
#'
#' @examples
#' # Get the UniProt entries of all proteins encoded by gene Pik3r1
#' ids <- c("P22682", "P47941")
#' query = list("accession_id" = ids)
#' df <-  query_uniprot(query = query)
#' head(df)
query_uniprot <- function(query = NULL,
                          base_url = "https://rest.uniprot.org/uniprotkb/",
                          columns = c("accession",
                                      "id",
                                      "gene_names",
                                      "organism_id",
                                      "reviewed"),
                          max_keys = 200,
                          updateProgress = NULL,
                          show_progress = TRUE) {

  if (max_keys > 200) {
    message("Parameter 'max_keys' exceeds 200.
            Try a lower value if the request fails.")
  }

  n_max <- max_keys / length(query)

  if (typeof(query) == "list") {

    idx_long <- which(sapply(query, length) > n_max)

    if (length(idx_long) == 1) {

      r <- length(query[[idx_long]]) %% n_max
      q <- length(query[[idx_long]]) %/% n_max

      if (r > 0) q <- q + 1

      df_list <- vector("list", length = q)

      if (show_progress) {
        cat("Querying the UniProt database...\n")
        pb <- utils::txtProgressBar(min = 0, max = q, style = 3)
      }

      for (i in 1:q) {

        i_start <- (i - 1) * n_max + 1
        i_end <- min(i * n_max, length(query[[idx_long]]))
        query_short <- query
        query_short[[idx_long]] <- query[[idx_long]][i_start : i_end]

        if (is.function(updateProgress)) {
          text <- paste(i_end, " / ", length(query[[idx_long]]), sep = "")
          updateProgress(value = as.numeric(format(i / q * 100, digits = 0)),
                         detail = text)
        }

        res <- get_uniprot_data(query = query_short,
                         base_url = base_url,
                         columns = columns)

        df_list[[i]] <- res$content

        if (show_progress) utils::setTxtProgressBar(pb, i)
      }

      if (show_progress) close(pb)

      return(do.call(rbind, df_list))

    }else {

      for (i in idx_long) {
        query_split <- vector("list", 2)

        query_split[[1]] <- query
        query_split[[2]] <- query

        query_split[[1]][[i]] <- query[[i]][1:n_max]
        query_split[[2]][[i]] <- query[[i]][(n_max + 1):length(query[[i]])]

        df_list <- lapply(query_split, query_uniprot,
                          base_url = base_url,
                          columns = columns,
                          max_keys = max_keys,
                          show_progress = show_progress)

        return(do.call(rbind, df_list))
      }
    }

  }

  res <- get_uniprot_data(query = query,
                          base_url = base_url,
                          columns = columns)

  return(res$content)

}


data-raw/generate_package_data.R
library(XML)

#### return_fields ####

path <- "data-raw/UniProtKB_return_fields_UniProt_help_UniProt.html"
tables <- XML::readHTMLTable(path, stringsAsFactors = FALSE)
data <- as.data.frame(do.call(rbind, tables))
rownames(data) <- NULL
data <- data[c(3, 1)]
names(data) <- gsub("Returned ", "", names(data), fixed = TRUE)
names(data) <- tolower(gsub("*", "", names(data), fixed = TRUE))
return_fields <- data
usethis::use_data(return_fields, overwrite = TRUE)

#### query_fields ####

path <- "data-raw/UniProtKB_query_fields_UniProt_help_UniProt.html"
tables <- XML::readHTMLTable(path, stringsAsFactors = FALSE)[[1]]
tables <- tables[-c(1, 2)]
names(tables) <- tolower(gsub("rest.uniprot.org ", "", names(tables)))
query_fields <- tables
usethis::use_data(query_fields, overwrite = TRUE)

#### uniprot_entries ####

devtools::load_all()
df <- query_uniprot(query = list("organism_id" = "10090", "reviewed" = "true"))
uniprot_entries <- df[1:1000, ]
usethis::use_data(uniprot_entries, overwrite = TRUE)


R/get_uniprot_data.R
#' Retrieve data from UniProt using UniProt's REST API
#'
#' @param query list of keys corresponding to UniProt's query fields.
#' For example :
#' list("gene_exact" = c("Pik3r1", "Pik3r2") ,
#' "organism" = c("10090", "9606"), "reviewed" = "yes").
#' See `query_fields` for available query fields.
#' @param base_url The base url for the UniProt REST API
#' @param columns names of UniProt data columns to retrieve.
#' Examples include "accession", "id", "gene_names", "keyword", "sequence".
#' See `return_fields` for available return fields.
#' @return a list with the following items :
#' \describe{
#'   \item{url}{the query url}
#'   \item{status}{the http status code for the request}
#'   \item{messages}{messages returned by the REST API}
#'   \item{content}{a data.frame containing the query results}
#' }
#' @import httr
#' @importFrom jsonlite fromJSON
#' @export
#'
#' @examples
#' #Getting gene names, keywords and protein sequences for a set of UniProt IDs.
#' ids <- c("P22682", "P47941")
#' cols <- c("accession", "id", "gene_names", "keyword", "sequence")
#' query = list("accession_id" = ids)
#' df <- get_uniprot_data(query = query, columns = cols)$content
#' df
get_uniprot_data <- function(query = NULL,
                             base_url = "https://rest.uniprot.org/uniprotkb/",
                             columns = c("accession",
                                         "id",
                                         "gene_names",
                                         "organism_id",
                                         "reviewed")) {

  full_url <- build_query_url(query = query,
                              base_url = base_url,
                              columns = columns,
                              format = "json")

  if (is.null(full_url)) return(NULL)

  # GET response to request

  resp <- try(httr::GET(full_url), silent = TRUE)

  if (inherits(resp, "try-error")){
    message(paste0("Request failed : could not get a response \n(",
                   resp[1], ")"))
    return(NULL)
  }

  # Check response type

  resp_type <- try(httr::http_type(resp), silent = TRUE)

  if (inherits(resp_type, "try-error")){
    message(paste0("Request failed : (", resp[1], ")"))
    return(NULL)
  }

  if (resp_type != "application/json") {
    message("Request failed : API did not return json")
    return(NULL)
  }

  # Parse the response

  content <- try(jsonlite::fromJSON(httr::content(resp,
                                        as = "text",
                                        encoding = "UTF-8"),
                                simplifyVector = FALSE),
                 silent = TRUE)


  if (! inherits(content, "list")){
    message("Request failed : could not read the response content")
    return(NULL)
  }

  messages <- unlist(content$messages)

  # check for invalid values and retry query without them

  df_invalid <- parse_messages(messages)

  if (!is.null(df_invalid)) {
    query <- clean_query(query, df_invalid)
    return(get_uniprot_data(query = query,
                            base_url = base_url,
                            columns = columns))
  }

  request_status <- httr::http_status(resp)
  request_status_code <- httr::status_code(resp)

  # request failure

  if(request_status_code != 200){

    # get message corresponding to request status

    http_message <- request_status$message

    # print an additional informative error message

      extra_message <- NULL
      n_query_items <- length(unlist(query))
      if (n_query_items > 200) {
        extra_message <- sprintf(
          "\nQuery has %s items and is probably too long.\n",
          n_query_items)
      }

      message(
        paste0(
          "\nUniProt API request failed : \n",
          paste(c(http_message, messages), collapse = "\n"),
          extra_message
        )
      )

    return(list(url = full_url,
                status = request_status_code,
                messages = c(http_message, messages),
                content = NULL))
  }

  # get query results and return them as a data.frame

  full_url <- build_query_url(query = query,
                              base_url = base_url,
                              columns = columns,
                              format = "tsv")

  resp <- try(httr::GET(full_url, ), silent = TRUE)

  if (inherits(resp, "try-error")){
    message(paste0("Request failed : could not get a response \n(",
                   resp[1], ")"))
    return(NULL)
  }

  # Check response type

  resp_type <- try(httr::http_type(resp), silent = TRUE)

  if (inherits(resp_type, "try-error")){
    message(paste0("Request failed : (", resp[1], ")"))
    return(NULL)
  }

  if (resp_type != "text/plain") {
    message("Request failed : API did not return plain text")
    return(NULL)
  }

  # Parse the response

  res <- try(httr::content(resp,
                           as = "text",
                           encoding = "UTF-8"),
             silent = TRUE)

  if (!inherits(res, "character")){
    message("Request failed : could not read the response content")
    return(NULL)
  }

  entries <- strsplit(res, split = "\n")[[1]]
  df <- as.data.frame(do.call(rbind,
                              lapply(entries,
                                     function(x) {
                                       strsplit(x, split = "\t")[[1]]
                                     })))
  names <- df[1, ]
  df <- as.data.frame(df[-1, ])
  names(df) <- names

  return(list(url = full_url,
              status = request_status_code,
              messages = messages,
              content = df))

}


#' Accessory function used to build the query url
#'
#' @param query list of keys corresponding to UniProt's query fields.
#' For example :
#' list("gene_exact" = c("Pik3r1", "Pik3r2") ,
#' "organism" = c("10090", "9606"), "reviewed" = "yes")
#' @param base_url The base url for the UniProt REST API
#' @param columns names of UniProt data columns to retrieve.
#' @param format format of the response provided by the UniProt API
#' @return the query url
build_query_url <- function(query = NULL,
                            base_url = "https://rest.uniprot.org/uniprotkb/",
                            columns = c("accession",
                                        "id",
                                        "gene_names",
                                        "organism_name",
                                        "reviewed"),
                            format = "json") {
  # format url from function arguments
  if (typeof(query) == "list") {
    formatted_queries <- sapply(seq_along(query),
                                function(x) {
                                  paste(names(query)[x],
                                        ":(",
                                        paste(query[[x]],
                                              collapse = "+OR+"),
                                        ")",
                                        sep = "")})

    full_query <- paste(formatted_queries, collapse = "+AND+")
  }else if (typeof(query) == "character" && length(query) == 1) {
    full_query <- query
  }else {
    message("Query not supported")
    return(NULL)
  }

  if (! "accession" %in% columns) {
    columns <- c("accession", columns)
  }

  cols <- paste(columns, collapse = ",")

  full_url <- paste(base_url,
                    "stream?query=",
                    full_query,
                    "&fields=", paste(cols, sep = ","),
                    "&format=", format,
                    sep = "")

  return(full_url)

}


#' Accessory function retrieving invalid values from messages returned by
#' the UniProt API.
#'
#' @param messages character string containing the error messages returned by
#' UniProt API
#' @return a data.frame with invalid values (in column "value") and
#' corresponding query field (in column "field"). NULL if no invalid values are
#' identified.
parse_messages <- function(messages) {

  pattern <- "^The '(.+)' filter value '(.+)' has invalid format"
  matches <- regexec(pattern, messages)
  m <- do.call(rbind, regmatches(messages, matches))

  if (length(m) > 0) {
    if (dim(m)[2] == 3) {
      df_invalid_values <- as.data.frame(m)
      names(df_invalid_values) <- c("message", "field", "value")
      return(df_invalid_values)
    }
  }

  return(NULL)
}



#' Accessory function removing invalid values from a query
#'
#' @param query list of keys corresponding to UniProt's query fields.
#' For example :
#' list("gene_exact" = c("Pik3r1", "Pik3r2") ,
#' "organism" = c("10090", "9606"), "reviewed" = "yes")
#' @param df data.frame with invalid values (in column "value") and
#' corresponding query field (in column "field").
#' @return the input query without the invalid values
clean_query <- function(query, df) {

  n_invalid <- dim(df)[1]

  message(paste0(n_invalid,
                 " invalid values were found (",
                 paste0(df$value[1 : min(10, n_invalid)],
                        collapse = ", "),
                 ifelse(n_invalid > 10, ", ...", ""),
                 ")",
                 " and removed from the query.")
  )

  fields <- unique(df$field)
  for (field in fields) {
    query[[field]] <- setdiff(query[[field]],
                              df$value[df$field == field])
    if (length(query[[field]]) == 0) {
      idx_field <- which(names(query) == field)
      query <- query[-idx_field]
      message(paste0("Field '",
                     field,
                     "' with no valid entries has been removed from query."))
    }
  }
  return(query)
}
